File: impl/tools/dsipvec/semantic.py
"""Stages 12–14: version negotiation, schema dispatch, stateless semantic checks.

Spec: §11, §10.3, §9.3, §13.2, §14.2, §15, §19.4; schema README checks 1, 2, 5, 7, 8, 9, 11.
"""
from __future__ import annotations

from typing import Any

from .registry import (resolve_reason, effective_answered_by, effective_progress_status, effective_integrity,
                       SUBSCRIPTION_EVENTS, REASON_BEARING_TYPES, INTEGRITY_MODES)
from .schema import schema_errors, dispatch_type
from .verdict import Verdict

INTRODUCTION_MAX_BYTES = 4096  # §19.4
DIRECTION_ANSWERS = {         # Impl (spec-gap 9): SDP-style offer→answer direction compatibility
    "sendrecv": {"sendrecv", "sendonly", "recvonly", "inactive"},
    "sendonly": {"recvonly", "inactive"},
    "recvonly": {"sendonly", "inactive"},
    "inactive": {"inactive"},
}


def _ver(s: str) -> tuple[int, int] | None:
    try:
        a, b = s.split(".")
        return int(a), int(b)
    except (ValueError, AttributeError):
        return None


def check_version(payload: dict, supported: dict) -> Verdict:
    """§11.2 compatibility rules. Malformed blocks are left to the schema stage."""
    blk = payload.get("dsip")
    if not isinstance(blk, dict):
        return Verdict.accept()
    core, min_core = _ver(blk.get("core")), _ver(blk.get("min_core"))
    mine = _ver(supported.get("core", "1.0"))
    if core is None or min_core is None or mine is None:
        return Verdict.accept()
    # Major versions are incompatible by default; the sender's floor must not exceed ours.
    if core[0] != mine[0] or min_core > mine:
        return Verdict.reject("version-unsupported", "session.unsupported-core-version")
    known = set(supported.get("profiles", [])) | set(supported.get("extensions", []))
    crit = blk.get("critical")
    if isinstance(crit, list):
        for c in crit:
            if c not in known:  # §11.2: unknown critical extensions require rejection
                return Verdict.reject("version-unsupported", "session.unsupported-critical-extension")
    profiles = blk.get("profiles")
    if isinstance(profiles, list) and profiles and not any(p in known for p in profiles):
        return Verdict.reject("version-unsupported", "session.unsupported-profile-version")
    return Verdict.accept()


def check_schema(payload: dict) -> Verdict:
    t = dispatch_type(payload)
    if t is None:
        return Verdict.reject("unknown-type")
    errs = schema_errors(t, payload)
    if errs:
        return Verdict.reject("schema-invalid", detail=errs[0][:200])
    return Verdict.accept()


def selection_is_subset(selection: dict, offer: dict) -> bool:
    """Check 9 (Impl, spec-gap 9)."""
    offered_media = offer.get("media", []) or []
    for sel in selection.get("media", []) or []:
        match = None
        for off in offered_media:
            if off.get("type") == sel.get("type") and off.get("purpose") == sel.get("purpose"):
                match = off
                break
        if match is None:
            return False
        offered_codecs = {c.get("id") for c in match.get("codecs", [])}
        if any(c.get("id") not in offered_codecs for c in sel.get("codecs", [])):
            return False
        if sel.get("direction") not in DIRECTION_ANSWERS.get(match.get("direction"), set()):
            return False
    offered_transports = {t.get("id") for t in offer.get("transports", []) or []}
    for t in selection.get("transports", []) or []:
        if t.get("id") not in offered_transports:
            return False
    return True


def check_semantic(payload: dict, ctx: dict, encoded_size: int | None = None) -> Verdict:
    """Stage 14. `ctx` is the raw vector context dict."""
    t = payload["type"]
    if t == "hello" and "in_reply_to" in payload and ctx.get("sent_hello_id") is not None:
        if payload["in_reply_to"] != ctx["sent_hello_id"]:  # §13.2 / §20.5 anti-splicing
            return Verdict.reject("hello-in-reply-to-mismatch")
    if t == "answer" and isinstance(ctx.get("offer"), dict):
        if not selection_is_subset(payload, ctx["offer"]):
            return Verdict.reject("selection-not-subset")
    if t == "subscribe":
        for ev in payload.get("events", []):
            cap = SUBSCRIPTION_EVENTS.get(ev)
            if cap is not None and payload["expires_in"] > cap:  # §9.3 hard caps → error policy.subscription-lifetime (v0.7)
                return Verdict.reject("subscription-lifetime-exceeded", "policy.subscription-lifetime")
    if t == "introduction":
        size = encoded_size if encoded_size is not None else ctx.get("encoded_size")
        if size is not None and size > INTRODUCTION_MAX_BYTES:  # §19.4
            return Verdict.reject("introduction-too-large")
    if t == "grant" and isinstance(ctx.get("known_introductions"), list):
        if payload["session"] not in ctx["known_introductions"]:  # §19.4
            return Verdict.reject("grant-unknown-introduction")
    if t == "key-rotation":  # §7.5 (v0.7, spec-gap 22)
        if payload["subject"] != payload["from"]:
            return Verdict.reject("rotation-subject-mismatch")
        if payload["next"] == payload["previous"]:
            return Verdict.reject("rotation-next-same-as-previous")
        signer = ctx.get("signer_kid")
        if signer is not None and not payload.get("recovery", False) and signer != payload["previous"]:
            return Verdict.reject("rotation-signer-not-previous")
    return Verdict.accept(**registry_effects(payload))


def registry_effects(payload: dict) -> dict:
    """Check 5 — membership with fallback. Never rejects."""
    t = payload["type"]
    eff: dict[str, Any] = {}
    warnings: list[str] = []
    if "reason" in payload and (t in REASON_BEARING_TYPES or t == "notify"):
        r = resolve_reason(payload["reason"], t)
        eff["reason"] = r.effective
        eff["fallback"] = r.fallback
        if not r.valid_on_type:
            warnings.append("reason-not-valid-on-type")
    if t in ("answer", "update") and "answered_by" in payload:
        eff["answered_by"] = effective_answered_by(payload["answered_by"])
    if t == "progress":
        eff["status"] = effective_progress_status(payload["status"])
    if t == "publish" and "integrity" in payload:
        eff["integrity"] = effective_integrity(payload["integrity"])
        if payload["integrity"] not in INTEGRITY_MODES:
            warnings.append("integrity-mode-unknown")  # §22.2 registry fallback
    out: dict[str, Any] = {}
    if eff:
        out["effective"] = eff
    if warnings:
        out["warnings"] = warnings
    return out


def check_payload(payload: dict, ctx: dict, encoded_size: int | None = None) -> Verdict:
    """Stages 12–14 in order."""
    for v in (check_version(payload, ctx.get("supported") or {"core": "1.0", "profiles": ["interactive-media/1.0"], "extensions": []}),
              check_schema(payload)):
        if not v.ok:
            return v
    return check_semantic(payload, ctx, encoded_size)

File: impl/crates/dsip-schema/src/semantic.rs
//! Stateless semantic checks (stages 12–14).
//!
//! Spec: §11 (version), §13.2/§20.5 (relay `hello` anti-splicing), §14.2
//! (selection ⊆ offer), §9.3 (subscription lifetime caps), §19.4
//! (introduction size, grant references an introduction), §15.1/§14.3/§12.10/§22.2
//! (registry membership with fallback), §7.5 (`key-rotation`: `from` = `subject`,
//! `next` ≠ `previous`, signer = `previous` unless `recovery`). Schema README checks 5, 7, 9, 11, 12.

use serde_json::{Map, Value};

use dsip_core::registry::{
    effective_answered_by, effective_integrity, effective_progress_status, resolve_reason, INTEGRITY_MODES,
    REASON_BEARING_TYPES, SUBSCRIPTION_EVENTS,
};
use dsip_core::version::{check_version, Supported};
use dsip_core::{RejectCode, Verdict, INTRODUCTION_MAX_BYTES};

use crate::validate::validate;

/// Receiver context for the stateless checks.
#[derive(Debug, Default, Clone)]
pub struct SemanticContext {
    /// Supported versions.
    pub supported: Supported,
    /// The id of the client `hello` this connection sent (relay `hello` anti-splicing).
    pub sent_hello_id: Option<String>,
    /// The offer (`media`, `transports`) an `answer` must select from.
    pub offer: Option<Value>,
    /// Pending introduction ids a `grant` may reference; `None` skips the check.
    pub known_introductions: Option<Vec<String>>,
    /// Encoded envelope size in bytes, when known (introduction cap).
    pub encoded_size: Option<usize>,
    /// The `kid` that signed the envelope, when known (`key-rotation` signer rule, §7.5).
    pub signer_kid: Option<String>,
}

impl SemanticContext {
    /// From a vector's `context` object.
    pub fn from_vector(ctx: &Value) -> SemanticContext {
        SemanticContext {
            supported: Supported::from_json(ctx.get("supported")),
            sent_hello_id: ctx.get("sent_hello_id").and_then(Value::as_str).map(String::from),
            offer: ctx.get("offer").filter(|o| o.is_object()).cloned(),
            known_introductions: ctx.get("known_introductions").and_then(Value::as_array).map(|a| {
                a.iter().filter_map(Value::as_str).map(String::from).collect()
            }),
            encoded_size: ctx.get("encoded_size").and_then(Value::as_u64).map(|n| n as usize),
            signer_kid: ctx.get("signer_kid").and_then(Value::as_str).map(String::from),
        }
    }
}

/// Impl (spec-gap 9): SDP-style offer→answer direction compatibility.
fn direction_answers(offered: &str, selected: &str) -> bool {
    matches!(
        (offered, selected),
        ("sendrecv", _) | ("sendonly", "recvonly" | "inactive") | ("recvonly", "sendonly" | "inactive") | ("inactive", "inactive")
    ) && ["sendrecv", "sendonly", "recvonly", "inactive"].contains(&selected)
}

/// Check 9: is `selection` (an `answer`) a subset of `offer`?
///
/// Spec: §14.2 — an answer is a selection, not a second offer. Impl (spec-gap 9)
/// for the per-field rule: match descriptors on `type`+`purpose`; codec ids ⊆
/// offered; direction answers the offered direction; transport id ∈ offered.
pub fn selection_is_subset(selection: &Value, offer: &Value) -> bool {
    let arr = |v: &Value, k: &str| v.get(k).and_then(Value::as_array).cloned().unwrap_or_default();
    let s = |v: &Value, k: &str| v.get(k).and_then(Value::as_str).map(String::from);
    let offered_media = arr(offer, "media");
    for sel in arr(selection, "media") {
        let Some(m) = offered_media.iter().find(|o| s(o, "type") == s(&sel, "type") && s(o, "purpose") == s(&sel, "purpose"))
        else {
            return false;
        };
        let offered_codecs: Vec<_> = arr(m, "codecs").iter().filter_map(|c| s(c, "id")).collect();
        if arr(&sel, "codecs").iter().any(|c| !s(c, "id").is_some_and(|id| offered_codecs.contains(&id))) {
            return false;
        }
        if !direction_answers(&s(m, "direction").unwrap_or_default(), &s(&sel, "direction").unwrap_or_default()) {
            return false;
        }
    }
    let offered_transports: Vec<_> = arr(offer, "transports").iter().filter_map(|t| s(t, "id")).collect();
    arr(selection, "transports").iter().all(|t| s(t, "id").is_some_and(|id| offered_transports.contains(&id)))
}

/// Stage 14: the stateless semantic checks, then registry effects on accept.
pub fn check_semantic(payload: &Value, ctx: &SemanticContext) -> Verdict {
    let t = payload.get("type").and_then(Value::as_str).unwrap_or("");
    let s = |k: &str| payload.get(k).and_then(Value::as_str);
    if t == "hello" {
        if let (Some(irt), Some(sent)) = (s("in_reply_to"), ctx.sent_hello_id.as_deref()) {
            if irt != sent {
                // §13.2 / §20.5: the relay hello is bound to the client hello it answers
                return Verdict::reject(RejectCode::HelloInReplyToMismatch);
            }
        }
    }
    if t == "answer" {
        if let Some(offer) = &ctx.offer {
            if !selection_is_subset(payload, offer) {
                return Verdict::reject(RejectCode::SelectionNotSubset);
            }
        }
    }
    if t == "subscribe" {
        let expires_in = payload.get("expires_in").and_then(Value::as_i64).unwrap_or(0);
        for ev in payload.get("events").and_then(Value::as_array).into_iter().flatten().filter_map(Value::as_str) {
            if let Some((_, cap)) = SUBSCRIPTION_EVENTS.iter().find(|(e, _)| *e == ev) {
                if expires_in > *cap {
                    // §9.3 hard caps → `error policy.subscription-lifetime` (v0.7, spec-gap 19)
                    return Verdict::reject_with(RejectCode::SubscriptionLifetimeExceeded, "policy.subscription-lifetime");
                }
            }
        }
    }
    if t == "introduction" {
        if let Some(size) = ctx.encoded_size {
            if size > INTRODUCTION_MAX_BYTES {
                return Verdict::reject(RejectCode::IntroductionTooLarge); // §19.4
            }
        }
    }
    if t == "grant" {
        if let Some(known) = &ctx.known_introductions {
            if !s("session").is_some_and(|sid| known.iter().any(|k| k == sid)) {
                return Verdict::reject(RejectCode::GrantUnknownIntroduction); // §19.4
            }
        }
    }
    if t == "key-rotation" {
        // §7.5 (v0.7, spec-gap 22): only the identity rotates its own keys, by its retiring key
        // unless a recovery key signs with `recovery: true`.
        if s("subject") != s("from") {
            return Verdict::reject(RejectCode::RotationSubjectMismatch);
        }
        if s("next") == s("previous") {
            return Verdict::reject(RejectCode::RotationNextSameAsPrevious);
        }
        let recovery = payload.get("recovery").and_then(Value::as_bool).unwrap_or(false);
        if let Some(signer) = ctx.signer_kid.as_deref() {
            if !recovery && Some(signer) != s("previous") {
                return Verdict::reject(RejectCode::RotationSignerNotPrevious);
            }
        }
    }
    let mut v = Verdict::accept();
    for (k, val) in registry_effects(payload) {
        v = v.with(&k, val);
    }
    v
}

/// Check 5 — registry membership with fallback. Never rejects; yields `effective` and `warnings`.
///
/// Spec: §15.1, §14.3, §12.10.
pub fn registry_effects(payload: &Value) -> Map<String, Value> {
    let t = payload.get("type").and_then(Value::as_str).unwrap_or("");
    let mut eff = Map::new();
    let mut warnings: Vec<Value> = vec![];
    if let Some(reason) = payload.get("reason").and_then(Value::as_str) {
        if REASON_BEARING_TYPES.contains(&t) || t == "notify" {
            let r = resolve_reason(reason, t);
            eff.insert("reason".into(), r.effective.into());
            eff.insert("fallback".into(), r.fallback.into());
            if !r.valid_on_type {
                warnings.push("reason-not-valid-on-type".into()); // Impl (spec-gap 10)
            }
        }
    }
    if (t == "answer" || t == "update") && payload.get("answered_by").is_some() {
        if let Some(a) = payload.get("answered_by").and_then(Value::as_str) {
            eff.insert("answered_by".into(), effective_answered_by(a).into());
        }
    }
    if t == "progress" {
        if let Some(st) = payload.get("status").and_then(Value::as_str) {
            eff.insert("status".into(), effective_progress_status(st).into());
        }
    }
    if t == "publish" {
        if let Some(m) = payload.get("integrity").and_then(Value::as_str) {
            eff.insert("integrity".into(), effective_integrity(Some(m)).into());
            if !INTEGRITY_MODES.contains(&m) {
                warnings.push("integrity-mode-unknown".into()); // §22.2 registry fallback
            }
        }
    }
    let mut out = Map::new();
    if !eff.is_empty() {
        out.insert("effective".into(), Value::Object(eff));
    }
    if !warnings.is_empty() {
        out.insert("warnings".into(), Value::Array(warnings));
    }
    out
}

/// Stages 12–14 in order: version, schema, semantic.
pub fn check_payload(payload: &Value, ctx: &SemanticContext) -> Verdict {
    let v = check_version(payload, &ctx.supported);
    if !v.ok() {
        return v;
    }
    let v = validate(payload);
    if !v.ok() {
        return v;
    }
    check_semantic(payload, ctx)
}
